cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuspatial/detail/iterator.hpp>
+#include <cuspatial/detail/utility/linestring.cuh>
+#include <cuspatial/experimental/detail/linestring_intersection_count.cuh>
+#include <cuspatial/experimental/geometry/segment.cuh>
+#include <cuspatial/experimental/ranges/range.cuh>
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <thrust/scan.h>
+#include <thrust/tuple.h>
+#include <thrust/uninitialized_fill.h>
+
+#include <utility>
+
+namespace cuspatial {
+namespace detail {
+
+/// Internal structure to provide convenient access to the intersection intermediate id arrays.
+template <typename IntegerRange>
+struct id_ranges {
+  using index_t = typename IntegerRange::value_type;
+  IntegerRange lhs_linestring_ids;
+  IntegerRange lhs_segment_ids;
+  IntegerRange rhs_linestring_ids;
+  IntegerRange rhs_segment_ids;
+
+  id_ranges(IntegerRange lhs_linestring_ids,
+            IntegerRange lhs_segment_ids,
+            IntegerRange rhs_linestring_ids,
+            IntegerRange rhs_segment_ids)
+    : lhs_linestring_ids(lhs_linestring_ids),
+      lhs_segment_ids(lhs_segment_ids),
+      rhs_linestring_ids(rhs_linestring_ids),
+      rhs_segment_ids(rhs_segment_ids)
+  {
+  }
```

Comment:
Is this something that should be written into another utility range object?

Reply:
No, this is something very specific to this internal function. `id_ranges` is just a convenience accessor for `intersection_intermediates` struct. You can represent this as a range of ranges, but readability I think what we have now is sufficient.